File: docs/beginner/module-02-basic-search/code-samples/python/06_result_processing.py
"""
Result Processing - Module 2 Python Examples
Processing and formatting search results from Azure AI Search

This module demonstrates:
- Basic result processing
- Result formatting for display
- Score analysis
- Result filtering and sorting
- Export capabilities
"""

import os
import sys
import logging
import json
from typing import List, Dict, Any, Optional
from dataclasses import dataclass

# Add setup directory to path for imports
sys.path.append(os.path.join(os.path.dirname(__file__), '..', '..', '..', '..', '..', 'setup'))
from connection_utils import get_default_search_client


@dataclass
class ProcessedResult:
    """Simple structure for processed search results"""
    title: str
    score: float
    author: str
    content_preview: str
    url: str
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary"""
        return {
            'title': self.title,
            'score': self.score,
            'author': self.author,
            'content_preview': self.content_preview,
            'url': self.url
        }


class ResultProcessor:
    """Class for processing search results"""
    
    def __init__(self, max_content_length: int = 150):
        self.max_content_length = max_content_length
        self.logger = logging.getLogger("result_processor")
    
    def process_raw_results(self, raw_results: List[Dict[str, Any]]) -> List[ProcessedResult]:
        """
        Convert raw search results to ProcessedResult objects
        
        Args:
            raw_results: List of raw result dictionaries
            
        Returns:
            List of ProcessedResult objects
        """
        processed_results = []
        
        for result in raw_results:
            try:
                # Extract and clean fields
                title = result.get('title', 'Untitled')
                score = result.get('@search.score', 0.0)
                author = result.get('author', 'Unknown')
                url = result.get('url', '#')
                
                # Create content preview
                content = result.get('content', '')
                content_preview = self._create_preview(content)
                
                processed_result = ProcessedResult(
                    title=title,
                    score=score,
                    author=author,
                    content_preview=content_preview,
                    url=url
                )
                
                processed_results.append(processed_result)
                
            except Exception as e:
                self.logger.error(f"Error processing result: {str(e)}")
                continue
        
        return processed_results
    
    def _create_preview(self, content: str) -> str:
        """Create a content preview with appropriate length"""
        if not content:
            return "No content available"
        
        if len(content) <= self.max_content_length:
            return content
        
        # Find a good breaking point
        preview = content[:self.max_content_length]
        last_space = preview.rfind(' ')
        
        if last_space > self.max_content_length * 0.8:
            preview = preview[:last_space]
        
        return preview + "..."
    
    def format_for_display(self, results: List[ProcessedResult]) -> str:
        """
        Format results for console display
        
        Args:
            results: List of ProcessedResult objects
            
        Returns:
            Formatted string for display
        """
        if not results:
            return "No results found."
        
        output = []
        output.append(f"\n{'='*60}")
        output.append(f"SEARCH RESULTS ({len(results)} found)")
        output.append(f"{'='*60}")
        
        for i, result in enumerate(results, 1):
            output.append(f"\n{i}. {result.title}")
            output.append(f"   Score: {result.score:.3f}")
            
            if result.author != "Unknown":
                output.append(f"   Author: {result.author}")
            
            if result.content_preview:
                output.append(f"   Preview: {result.content_preview}")
            
            if result.url != "#":
                output.append(f"   URL: {result.url}")
            
            output.append(f"   {'-'*50}")
        
        return "\n".join(output)
    
    def sort_by_score(self, results: List[ProcessedResult], reverse: bool = True) -> List[ProcessedResult]:
        """Sort results by score"""
        return sorted(results, key=lambda x: x.score, reverse=reverse)
    
    def filter_by_score(self, results: List[ProcessedResult], min_score: float) -> List[ProcessedResult]:
        """Filter results by minimum score"""
        return [result for result in results if result.score >= min_score]
    
    def analyze_scores(self, results: List[ProcessedResult]) -> Dict[str, float]:
        """Analyze score distribution"""
        if not results:
            return {}
        
        scores = [result.score for result in results]
        
        return {
            'count': len(scores),
            'min_score': min(scores),
            'max_score': max(scores),
            'avg_score': sum(scores) / len(scores),
            'score_range': max(scores) - min(scores)
        }
    
    def export_to_json(self, results: List[ProcessedResult], filename: str):
        """Export results to JSON file"""
        try:
            results_dict = [result.to_dict() for result in results]
            
            with open(filename, 'w', encoding='utf-8') as f:
                json.dump(results_dict, f, indent=2, ensure_ascii=False)
            
            print(f"✅ Results exported to {filename}")
            
        except Exception as e:
            print(f"❌ Error exporting to JSON: {str(e)}")


def demonstrate_result_processing():
    """Demonstrate result processing capabilities"""
    print("🔧 Result Processing Demonstration")
    print("=" * 50)
    
    try:
        # Get some sample results
        search_client = get_default_search_client()
        raw_results = search_client.search(search_text="python programming", top=5)
        raw_result_list = list(raw_results)
        
        if not raw_result_list:
            print("❌ No results found for demo. Make sure your index has data.")
            return
        
        # Process results
        processor = ResultProcessor()
        processed_results = processor.process_raw_results(raw_result_list)
        
        print(f"✅ Processed {len(processed_results)} results")
        
        # Display formatted results
        formatted_output = processor.format_for_display(processed_results[:3])
        print(formatted_output)
        
        # Analyze scores
        stats = processor.analyze_scores(processed_results)
        print(f"\n📊 Score Analysis:")
        print(f"   Total results: {stats['count']}")
        print(f"   Score range: {stats['min_score']:.3f} - {stats['max_score']:.3f}")
        print(f"   Average score: {stats['avg_score']:.3f}")
        
        # Filter high-quality results
        high_quality = processor.filter_by_score(processed_results, 1.0)
        print(f"\n🎯 High-quality results (score ≥ 1.0): {len(high_quality)}")
        
        # Export to JSON
        processor.export_to_json(processed_results, "sample_results.json")
        
        print("\n✅ Result processing demonstration completed!")
        
    except Exception as e:
        print(f"❌ Demo failed: {str(e)}")


if __name__ == "__main__":
    # Setup logging
    logging.basicConfig(level=logging.INFO, format='%(levelname)s: %(message)s')
    
    # Run demonstration
    demonstrate_result_processing()
    
    print("\n💡 Next Steps:")
    print("   - Try processing results from different searches")
    print("   - Experiment with different filtering criteria")
    print("   - Check out 07_error_handling.py for robust error handling")
    print("   - Learn about search patterns in 08_search_patterns.py")

File: docs/beginner/module-02-basic-search/code-samples/javascript/06_result_processing.js
/**
 * Result Processing - Module 2 JavaScript Examples
 * Processing and formatting search results from Azure AI Search
 * 
 * This module demonstrates:
 * - Basic result processing
 * - Result formatting for display
 * - Score analysis
 * - Result filtering and sorting
 * - Export capabilities
 */

const { SearchClient, AzureKeyCredential } = require('@azure/search-documents');
const fs = require('fs').promises;

/**
 * Simple structure for processed search results
 */
class ProcessedResult {
    constructor(title, score, author, contentPreview, url) {
        this.title = title || 'Untitled';
        this.score = score || 0.0;
        this.author = author || 'Unknown';
        this.contentPreview = contentPreview || 'No content available';
        this.url = url || '#';
    }

    /**
     * Convert to plain object
     * @returns {Object} Plain object representation
     */
    toObject() {
        return {
            title: this.title,
            score: this.score,
            author: this.author,
            contentPreview: this.contentPreview,
            url: this.url
        };
    }
}

/**
 * Class for processing search results
 */
class ResultProcessor {
    /**
     * Initialize the result processor
     * @param {number} maxContentLength - Maximum length for content previews
     */
    constructor(maxContentLength = 150) {
        this.maxContentLength = maxContentLength;
    }

    /**
     * Convert raw search results to ProcessedResult objects
     * @param {Array} rawResults - Array of raw search results
     * @returns {Array<ProcessedResult>} Array of ProcessedResult objects
     */
    processRawResults(rawResults) {
        const processedResults = [];

        for (const result of rawResults) {
            try {
                const document = result.document || {};

                // Extract and clean fields
                const title = document.title || 'Untitled';
                const score = result.score || 0.0;
                const author = document.author || 'Unknown';
                const url = document.url || '#';

                // Create content preview
                const content = document.content || '';
                const contentPreview = this.createPreview(content);

                const processedResult = new ProcessedResult(
                    title, score, author, contentPreview, url
                );

                processedResults.push(processedResult);

            } catch (error) {
                console.error(`Error processing result: ${error.message}`);
                continue;
            }
        }

        return processedResults;
    }

    /**
     * Create a content preview with appropriate length
     * @param {string} content - Full content text
     * @returns {string} Preview text
     */
    createPreview(content) {
        if (!content) {
            return 'No content available';
        }

        if (content.length <= this.maxContentLength) {
            return content;
        }

        // Find a good breaking point
        let preview = content.substring(0, this.maxContentLength);
        const lastSpace = preview.lastIndexOf(' ');

        if (lastSpace > this.maxContentLength * 0.8) {
            preview = preview.substring(0, lastSpace);
        }

        return preview + '...';
    }

    /**
     * Format results for console display
     * @param {Array<ProcessedResult>} results - Array of ProcessedResult objects
     * @returns {string} Formatted string for display
     */
    formatForDisplay(results) {
        if (!results || results.length === 0) {
            return 'No results found.';
        }

        const output = [];
        output.push('');
        output.push('='.repeat(60));
        output.push(`SEARCH RESULTS (${results.length} found)`);
        output.push('='.repeat(60));

        results.forEach((result, index) => {
            output.push(`\n${index + 1}. ${result.title}`);
            output.push(`   Score: ${result.score.toFixed(3)}`);

            if (result.author !== 'Unknown') {
                output.push(`   Author: ${result.author}`);
            }

            if (result.contentPreview && result.contentPreview !== 'No content available') {
                output.push(`   Preview: ${result.contentPreview}`);
            }

            if (result.url !== '#') {
                output.push(`   URL: ${result.url}`);
            }

            output.push(`   ${'-'.repeat(50)}`);
        });

        return output.join('\n');
    }

    /**
     * Sort results by score
     * @param {Array<ProcessedResult>} results - Results to sort
     * @param {boolean} descending - Sort in descending order
     * @returns {Array<ProcessedResult>} Sorted results
     */
    sortByScore(results, descending = true) {
        return [...results].sort((a, b) => {
            return descending ? b.score - a.score : a.score - b.score;
        });
    }

    /**
     * Filter results by minimum score
     * @param {Array<ProcessedResult>} results - Results to filter
     * @param {number} minScore - Minimum score threshold
     * @returns {Array<ProcessedResult>} Filtered results
     */
    filterByScore(results, minScore) {
        return results.filter(result => result.score >= minScore);
    }

    /**
     * Analyze score distribution
     * @param {Array<ProcessedResult>} results - Results to analyze
     * @returns {Object} Score statistics
     */
    analyzeScores(results) {
        if (!results || results.length === 0) {
            return {};
        }

        const scores = results.map(result => result.score);

        const minScore = Math.min(...scores);
        const maxScore = Math.max(...scores);
        const avgScore = scores.reduce((sum, score) => sum + score, 0) / scores.length;

        return {
            count: scores.length,
            minScore: minScore,
            maxScore: maxScore,
            avgScore: avgScore,
            scoreRange: maxScore - minScore
        };
    }

    /**
     * Export results to JSON file
     * @param {Array<ProcessedResult>} results - Results to export
     * @param {string} filename - Output filename
     */
    async exportToJson(results, filename) {
        try {
            const resultsArray = results.map(result => result.toObject());
            const json = JSON.stringify(resultsArray, null, 2);
            
            await fs.writeFile(filename, json, 'utf8');
            console.log(`✅ Results exported to ${filename}`);

        } catch (error) {
            console.error(`❌ Error exporting to JSON: ${error.message}`);
        }
    }
}

/**
 * Demonstrate result processing capabilities
 */
async function demonstrateResultProcessing() {
    console.log('🔧 Result Processing Demonstration');
    console.log('='.repeat(50));

    try {
        // Initialize search client (replace with your actual service details)
        const searchClient = new SearchClient(
            'https://your-service.search.windows.net',
            'your-index-name',
            new AzureKeyCredential('your-api-key')
        );

        // Get some sample results
        const searchOptions = { top: 5 };
        const rawResults = await searchClient.search('python programming', searchOptions);

        const rawResultArray = [];
        for await (const result of rawResults.results) {
            rawResultArray.push(result);
        }

        if (rawResultArray.length === 0) {
            console.log('❌ No results found for demo. Make sure your index has data.');
            return;
        }

        // Process results
        const processor = new ResultProcessor();
        const processedResults = processor.processRawResults(rawResultArray);

        console.log(`✅ Processed ${processedResults.length} results`);

        // Display formatted results
        const formattedOutput = processor.formatForDisplay(processedResults.slice(0, 3));
        console.log(formattedOutput);

        // Analyze scores
        const stats = processor.analyzeScores(processedResults);
        console.log('\n📊 Score Analysis:');
        console.log(`   Total results: ${stats.count}`);
        console.log(`   Score range: ${stats.minScore.toFixed(3)} - ${stats.maxScore.toFixed(3)}`);
        console.log(`   Average score: ${stats.avgScore.toFixed(3)}`);

        // Filter high-quality results
        const highQuality = processor.filterByScore(processedResults, 1.0);
        console.log(`\n🎯 High-quality results (score ≥ 1.0): ${highQuality.length}`);

        // Export to JSON
        await processor.exportToJson(processedResults, 'sample_results.json');

        console.log('\n✅ Result processing demonstration completed!');

    } catch (error) {
        console.error(`❌ Demo failed: ${error.message}`);
    }
}

/**
 * Show result processing best practices
 */
function resultProcessingBestPractices() {
    console.log('\n📚 Result Processing Best Practices');
    console.log('='.repeat(50));

    console.log('\n💡 Processing Guidelines:');
    console.log('   ✅ Always handle missing or null fields gracefully');
    console.log('   ✅ Create meaningful content previews');
    console.log('   ✅ Normalize and clean data consistently');
    console.log('   ✅ Provide fallback values for missing data');

    console.log('\n📊 Score Analysis Tips:');
    console.log('   ✅ Understand your score distribution');
    console.log('   ✅ Set appropriate quality thresholds');
    console.log('   ✅ Monitor score patterns over time');
    console.log('   ✅ Use scores to improve search relevance');

    console.log('\n🎨 Display Formatting:');
    console.log('   ✅ Keep previews concise but informative');
    console.log('   ✅ Highlight important information');
    console.log('   ✅ Provide consistent formatting');
    console.log('   ✅ Make results scannable');

    console.log('\n💾 Export Considerations:');
    console.log('   ✅ Choose appropriate export formats');
    console.log('   ✅ Handle large result sets efficiently');
    console.log('   ✅ Include metadata in exports');
    console.log('   ✅ Validate exported data');
}

// Main execution
async function main() {
    try {
        await demonstrateResultProcessing();
        resultProcessingBestPractices();

        console.log('\n💡 Next Steps:');
        console.log('   - Try processing results from different searches');
        console.log('   - Experiment with different filtering criteria');
        console.log('   - Check out 07_error_handling.js for robust error handling');
        console.log('   - Learn about search patterns in 08_search_patterns.js');
    } catch (error) {
        console.error(`❌ Main execution failed: ${error.message}`);
    }
}

// Export for use in other modules
module.exports = {
    ProcessedResult,
    ResultProcessor,
    demonstrateResultProcessing,
    resultProcessingBestPractices
};

// Run if this file is executed directly
if (require.main === module) {
    main();
}